src/main/java/org/springframework/retry/RetryOperations.java

```diff
@@ -76,6 +116,34 @@ <T, E extends Throwable> T execute(RetryCallback<T, E> retryCallback, RecoveryCa
 	<T, E extends Throwable> T execute(RetryCallback<T, E> retryCallback, RetryState retryState)
 			throws E, ExhaustedRetryException;
 
+	/**
+	 * A simple stateful retry. Execute the supplied {@link RetryCallback} with a target
+	 * object for the attempt identified by the {@link DefaultRetryState}. Exceptions
+	 * thrown by the callback are always propagated immediately so the state is required
+	 * to be able to identify the previous attempt, if there is one - hence the state is
+	 * required. Normal patterns would see this method being used inside a transaction,
+	 * where the callback might invalidate the transaction if it fails.
+	 *
+	 * See implementations for configuration details.
+	 * @param retryCallback the {@link RetryCallback}
+	 * @param retryState the {@link RetryState}
+	 * @param <T> the type of the return value
+	 * @param <E> the type of the exception to return
+	 * @return the value returned by the {@link RetryCallback} upon successful invocation,
+	 * and that returned by the {@link RecoveryCallback} otherwise.
+	 * @param contextKey a key to use to map the context when
+	 * {@link RetrySynchronizationManager#isUseThreadLocal()} is false. Must be unique for
+	 * each execution.
```

Comment:
Cannot we have it randomly generated then instead?
Or something like `contextKeySupplier` injected into a `RetryTemplate`...

Reply:
Good idea.